Add kutil_urldecode()

Aversiste opened the pull request:
As expected this function does the inverse of kutil_urlencode().

I added the short documentation for this function in `kutil_urlencode.3` and recycled the regress tests in `test-urlencode.c`, with an additional one for invalid triplets like %zu which are not base 16 integers. I use this function in a kcgi-based project where I need to work with url-encoded strings and thought it could be useful for others too.

What do you think ?

kristapsdz commented:
There's already a urldecode() function in child.c.  Can you modify your patch to use it?  You can do so by copying the urldecode() function into kcgi.c and renaming it as kcgi_urldecode_inplace (or whichever you prefer---it's yours!) and making it public.  Then add kcgi_urldecode that basically strdup's the input variable before passing it to the new function.

However, the kcgi_urldecode function shouldn't return a char pointer.  It should return a kcgi_err and set the return as a double-pointer.

```c
enum kcgi_err kcgi_urldecode(const char *, char **);
enum kcgi_err kcgi_urldecode_inplace(char *);
```

This would then be able to return KCGI_FORM on bad format and KCGI_ENOMEM on memory allocation failure.

Aversiste commented:
Sorry, I never noticed there was already an implementation in the tree. I reworked the patch yesterday to use it instead and updated the man page too. However I kept the name as kutil_urldecode() to match the naming of kutil_urlencode(), hope it is okay for you.

kristapsdz commented:
All checked in!  I clarified a bit of corner case behaviour, added some tests, and other misc.  Please reopen if you find issues.  Thank you!

dpgao commented:
Sorry for reopening this but I believe that the `kutil_urldecode_inplace` function introduced here runs in quatratic time in the worst case because of the linear `strlen` and `memmov` calls in the loop. This can be optimised to run in linear time by using another pointer to keep track of the undecoded portion of `p`, which will be moved to the front only when the decode fails.
